### ISSUE TITLE
cpptest does not work

### ISSUE THREAD

**Issue by anacierdem** [DragonMinded/libdragon#92]
It compiles OK but linking with `g++` does not work.
See the Makefile -> `LD = $(GCCN64PREFIX)g++`

**Comment by SmithersPrime** [comment]
I'm running into a different issue with this example, I'm able to build this but it doesn't run. Other examples like the sprite test work fine.

**Comment by command-tab** [comment]
I'm able to successfully build cpptest with libdragon-docker, but when I run it in cen64, the screen remains white.

**Comment by anacierdem** [comment]
I mean the linking step produces a broken rom image. We need to disassemble the binary to see why it is not working as expected. When we were linking with ld, it was working but c++ features were not linking/working properly as expected.

**Comment by bryanperris** [comment]
This is because the ` __CTOR_LIST__ ` pointer points to an address with the value of 0xFFFFFFFF

In the __do_global_ctors function, MIPS first does a jump to 0xFFFFFFFF via the JALR instruction which will crash the system.

I was wondering why this test didn't work in my emulator, and then I used Ghidra to analyze the ELF file

```
                             __CTOR_LIST__                                   XREF[2]:     Entry Point(*), 
                                                                                          __do_global_ctors:80002838(R)  
        8002871c ff ff ff ff     undefined4 FFFFFFFFh
```

If you were to add 4 to the 8002871C address, then it will hit the valid constructor function pointer.

```
                             PTR__GLOBAL__sub_I_o1_80028720                  XREF[2]:     __do_global_ctors:80002838(R), 
                                                                                          __do_global_ctors_aux:80021330(R
        80028720 80 02 14 10     addr       _GLOBAL__sub_I_o1
```

Here the constructor function dump:

```
                             **************************************************************
                             *                          FUNCTION                          *
                             **************************************************************
                             undefined _GLOBAL__sub_I_o1()
                               assume gp = 0x80031110
             undefined         v0_lo:1        <RETURN>
                             _GLOBAL__sub_I_o1                               XREF[4]:     __do_global_ctors:8000283c(c), 
                                                                                          __do_global_ctors_aux:80021328(c
                                                                                          __do_global_ctors_aux:80021330(*
                                                                                          80028720(*)  
        80021410 24 02 00 02     li         v0,0x2
             assume gp = <UNKNOWN>
        80021414 03 e0 00 08     jr         ra
        80021418 af 82 80 a0     _sw        v0,-0x7f60(gp)=>o1                               = ??
        8002141c 00              ??         00h
        8002141d 00              ??         00h
        8002141e 00              ??         00h
        8002141f 00              ??         00h
```

**Comment by anacierdem** [comment]
Did forget about the reference on that PR. It is not actually fixing all the problems, but it is a good start.

**Comment by meeq** [comment]
@anacierdem I just tested the fix from #118 on real hardware using an EverDrive 2.0, and it appears to work just fine.

[cpptest.z64.zip](https://github.com/DragonMinded/libdragon/files/5877335/cpptest.z64.zip)

**Comment by anacierdem** [comment]
Interesting. Does not seem to work on v3. Let me try again if it is a contact issue or something.

**Comment by anacierdem** [comment]
It works fine, I probably did forget something.
